File: csharp/robot-name/RobotName.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Robot
{
    private static readonly HashSet<string> names = new HashSet<string>();

    private static readonly char[] alphabet = Enumerable
        .Range('A', 'Z' - 'A' + 1)
        .Select(letter => (Char)letter)
        .ToArray();

    private static Random random = new Random();

    private char RandomLetter => alphabet[random.Next(26)];

    private string RandomNumber => random.Next(1000).ToString("000");

    public string Name { get; private set; }

    public Robot() => Name = GenerateName();

    public void Reset() => Name = GenerateName();

    private string GenerateName()
    {
        var name = $"{RandomLetter}{RandomLetter}{RandomNumber}";

        if (names.Add(name))
            return name;
        else
            return GenerateName();
    }
}

File: fsharp/robot-name/RobotName.fs
module RobotName
open System

type Robot = { name: string }

let alphabet = ['A' .. 'Z']

let random n = Random().Next(n)

let getRandomLetter() =
    alphabet.[random 26]

let getRandomName() =
    let letters = [getRandomLetter(); getRandomLetter()] |> List.map string
    String.concat "" <| letters

let getRandomNumber() =
    let number = random 1000
    number.ToString("000")

let mkRobot() =
    { name = getRandomName() + getRandomNumber() }

let name robot = robot.name

let reset robot = mkRobot()